File: blockchain.py
import hashlib
import json
import time
from block import Block

# --- QTP BLOCKCHAIN ---
# This is the chain itself — the living, growing record of every
# transaction that has ever happened on the QTP network

class Blockchain:
    def __init__(self):
        # Mining difficulty — how many leading zeros a valid hash needs
        # 4 is low for testing, real QTP mainnet will be much higher
        self.difficulty     = 4

        # The mempool holds transactions waiting to be included in a block
        # When a miner mines a block, they pull from here
        self.mempool        = []

        # Mining reward — exactly as specified in the QTP whitepaper
        # 50 QTP per block, halving every 210,000 blocks
        self.block_reward   = 50.0

        # The chain itself — starts with the genesis block
        self.chain          = [self.create_genesis_block()]

    def create_genesis_block(self):
        # The genesis block is the first block — hardcoded into the protocol
        # It has no previous block so previous_hash is all zeros
        # This is block 0 of the QTP blockchain
        print("Creating QTP genesis block...")
        genesis = Block(
            index=         0,
            transactions=  [{
                "sender":    "GENESIS",
                "recipient": "GENESIS",
                "amount":    0,
                "note":      "Qubit TopCoin — For everyone. Forever."
            }],
            previous_hash= "0" * 64
        )
        genesis.mine(self.difficulty)
        return genesis

    def get_latest_block(self):
        # The tip of the chain — the most recently mined block
        return self.chain[-1]

    def get_balance(self, address):
        # Calculate a wallet's balance by scanning every transaction
        # on the entire chain — just like Bitcoin does it
        balance = 0.0

        for block in self.chain:
            for tx in block.transactions:
                # Receiving QTP increases balance
                if tx["recipient"] == address:
                    balance += tx["amount"]
                # Sending QTP decreases balance
                if tx["sender"] == address:
                    balance -= tx["amount"]

        return balance

    def add_transaction(self, transaction):
        # Add a pending transaction to the mempool
        # In a full implementation we would verify the signature here
        # We will add that in the next step
        self.mempool.append(transaction)
        print(f"  Transaction added to mempool. "
              f"Pending transactions: {len(self.mempool)}")

    def mine_pending_transactions(self, miner_address):
        # This is what miners do — they take pending transactions
        # from the mempool, bundle them into a block, and mine it

        # Step 1: Create the mining reward transaction
        # This is how new QTP enters circulation — paid to the miner
        # The halving schedule from the whitepaper is enforced here
        current_reward = self.get_current_reward()
        reward_tx = {
            "sender":    "NETWORK",
            "recipient": miner_address,
            "amount":    current_reward,
            "note":      f"Block {len(self.chain)} mining reward"
        }

        # Step 2: Bundle reward + mempool transactions into a new block
        transactions = [reward_tx] + self.mempool

        new_block = Block(
            index=         len(self.chain),
            transactions=  transactions,
            previous_hash= self.get_latest_block().hash
        )

        # Step 3: Mine the block — this is the proof of work
        print(f"\nMining block {new_block.index}...")
        new_block.mine(self.difficulty)

        # Step 4: Add the mined block to the chain
        self.chain.append(new_block)

        # Step 5: Clear the mempool — those transactions are now confirmed
        self.mempool = []

        print(f"  Block {new_block.index} added to chain.")
        print(f"  Miner {miner_address[:20]}... earned {current_reward} QTP")
        return new_block

    def get_current_reward(self):
        # QTP halving schedule — exactly as in the whitepaper
        # 50 QTP to start, halving every 210,000 blocks
        halvings = len(self.chain) // 210000
        reward   = self.block_reward / (2 ** halvings)
        return reward

    def is_chain_valid(self):
        # Validate the entire chain from block 1 to present
        # This is what every QTP node does when it joins the network
        print("\nValidating entire blockchain...")

        for i in range(1, len(self.chain)):
            current  = self.chain[i]
            previous = self.chain[i - 1]

            # Rule 1: The block's stored hash must match a fresh calculation
            if not current.is_valid():
                print(f"  INVALID: Block {i} hash is corrupted")
                return False

            # Rule 2: This block must point to the previous block's hash
            if current.previous_hash != previous.hash:
                print(f"  INVALID: Block {i} is disconnected from chain")
                return False

        print(f"  Chain valid — {len(self.chain)} blocks verified")
        return True


# --- TEST THE FULL BLOCKCHAIN ---
print("=" * 55)
print("  QTP BLOCKCHAIN TEST")
print("=" * 55)
print("")

# Start the blockchain — this mines the genesis block
qtpchain = Blockchain()
print("")

# Two test wallet addresses
miner_address  = "GbtE3vPeomLSRNN2k9DsDfy8YL164JXMw2"
sender_address = "GMVekpeEgiMKED7bQVCVp6nqb8Q597k2FQ"

# Add some transactions to the mempool
print("Adding transactions to mempool...")
qtpchain.add_transaction({
    "sender":    miner_address,
    "recipient": sender_address,
    "amount":    12.5
})
qtpchain.add_transaction({
    "sender":    sender_address,
    "recipient": miner_address,
    "amount":    3.0
})

# Mine the pending transactions
qtpchain.mine_pending_transactions(miner_address)

# Check balances
print(f"\nBalance of miner:  {qtpchain.get_balance(miner_address)} QTP")
print(f"Balance of sender: {qtpchain.get_balance(sender_address)} QTP")

# Validate the chain
valid = qtpchain.is_chain_valid()
print(f"Chain valid: {valid}")

# Now tamper with the chain and prove it gets caught
print("\nTampering with block 1 transaction...")
qtpchain.chain[1].transactions[0]["amount"] = 999999.0
valid_after_tamper = qtpchain.is_chain_valid()
print(f"Chain valid after tampering: {valid_after_tamper}")

print("")
print("=" * 55)
print("  BLOCKCHAIN COMPLETE")
print("=" * 55)

File: qtpcore/src/blockchain.rs
use crate::block::Block;
use crate::transaction::{SignedTransaction, TransactionData};
use crate::wallet::Wallet;

const INITIAL_REWARD: u64      = 5_000_000_000;
const HALVING_INTERVAL: u64    = 210_000;
const TARGET_BLOCK_TIME: u64   = 150;
const ADJUSTMENT_INTERVAL: u64 = 10;

pub struct Blockchain {
    pub chain:      Vec<Block>,
    pub mempool:    Vec<SignedTransaction>,
    pub difficulty: usize,
}

impl Blockchain {
    pub fn new() -> Self {
        let difficulty = 4;
        println!("Initializing QTP blockchain...");
        let genesis = Block::genesis(difficulty);
        Blockchain {
            chain:      vec![genesis],
            mempool:    vec![],
            difficulty,
        }
    }

    pub fn latest_block(&self) -> &Block {
        self.chain.last().unwrap()
    }

    pub fn height(&self) -> u64 {
        self.chain.len() as u64
    }

    pub fn add_transaction(&mut self, tx: SignedTransaction) {
        self.mempool.push(tx);
        self.mempool.sort_by(|a, b| b.data.fee.cmp(&a.data.fee));
        println!(
            "  Transaction added to mempool. Pending: {}",
            self.mempool.len()
        );
    }

    pub fn get_balance(&self, address: &str) -> u64 {
        let mut received: u64 = 0;
        let mut sent: u64     = 0;
        for block in &self.chain {
            for tx in &block.transactions {
                if tx.data.recipient == address {
                    received = received.saturating_add(tx.data.amount);
                }
                if tx.data.sender == address {
                    sent = sent.saturating_add(tx.data.amount);
                    sent = sent.saturating_add(tx.data.fee);
                }
            }
        }
        received.saturating_sub(sent)
    }

    pub fn get_current_reward(&self) -> u64 {
        let halvings = self.height() / HALVING_INTERVAL;
        if halvings >= 64 { return 0; }
        INITIAL_REWARD >> halvings
    }

    pub fn get_total_fees(&self, transactions: &[SignedTransaction]) -> u64 {
        transactions.iter().map(|tx| tx.data.fee).sum()
    }

    pub fn adjust_difficulty(&mut self) {
        let height = self.height();
        if height % ADJUSTMENT_INTERVAL != 0 || height < ADJUSTMENT_INTERVAL {
            return;
        }
        let last_block  = self.latest_block();
        let first_block = &self.chain[(height - ADJUSTMENT_INTERVAL) as usize];
        let actual_time   = last_block.timestamp - first_block.timestamp;
        let expected_time = TARGET_BLOCK_TIME * ADJUSTMENT_INTERVAL;
        if actual_time < expected_time / 2 {
            self.difficulty += 1;
            println!("  Difficulty increased to {} (too fast: {}s vs {}s)",
                self.difficulty, actual_time, expected_time);
        } else if actual_time > expected_time * 2 && self.difficulty > 1 {
            self.difficulty -= 1;
            println!("  Difficulty decreased to {} (too slow: {}s vs {}s)",
                self.difficulty, actual_time, expected_time);
        } else {
            println!("  Difficulty unchanged at {} ({}s vs {}s expected)",
                self.difficulty, actual_time, expected_time);
        }
    }

    pub fn mine_pending_transactions(
        &mut self,
        miner_address: &str,
        miner_wallet:  &Wallet,
    ) {
        self.adjust_difficulty();

        let reward       = self.get_current_reward();
        let fees         = self.get_total_fees(&self.mempool);
        let total_payout = reward + fees;

        let reward_data = TransactionData {
            sender:    "NETWORK".to_string(),
            recipient: miner_address.to_string(),
            amount:    total_payout,
            fee:       0,
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        };

        let signed_reward = crate::transaction::sign_transaction(
            &reward_data,
            &miner_wallet.secret_key,
            &miner_wallet.public_key,
        ).expect("Reward signing failed");

        let mut transactions = vec![signed_reward];
        transactions.extend(self.mempool.drain(..));

        let mut new_block = Block::new(
            self.height(),
            transactions,
            self.latest_block().hash.clone(),
        );

        println!("\nMining block {} at difficulty {}...",
            new_block.index, self.difficulty);
        new_block.mine(self.difficulty);

        println!("  Miner earned {} cori reward + {} cori fees = {} cori total",
            reward, fees, total_payout);

        self.chain.push(new_block);
    }

    pub fn is_valid(&self) -> bool {
        println!("\nValidating blockchain...");
        for i in 1..self.chain.len() {
            let current  = &self.chain[i];
            let previous = &self.chain[i - 1];
            if !current.is_valid() {
                println!("  INVALID: Block {} hash corrupted", i);
                return false;
            }
            if current.previous_hash != previous.hash {
                println!("  INVALID: Block {} disconnected from chain", i);
                return false;
            }
        }
        println!("  Chain valid — {} blocks verified", self.chain.len());
        true
    }
}
